launcher/modManager/extractor/hdextractor.cpp

```diff
@@ -0,0 +1,648 @@
+
+	if(file.open(QIODevice::ReadOnly))
+	{
+		QDataStream in(&file);
+		in.setByteOrder(QDataStream::LittleEndian);
+		
+		int32_t infoOffset;
+		int32_t fileCount;
+
+		file.seek(file.pos() + 4); // dummy
+		in >> infoOffset;
+		file.seek(infoOffset);
+		in >> fileCount;
+
+		int32_t nameOffset = file.pos();
+		for(int i = 0; i < fileCount; i++)
+		{
+			QString name;
+			int32_t offset;
+			int32_t imageConfigSize;
+			int32_t chunks;
+			int32_t zsize;
+			int32_t size;
+			QString imageConfig;
+			QByteArray data;
+			QByteArray dataCompressed;
+			std::vector<QByteArray> dataCompressedArray;
+
+			file.seek(nameOffset);
+			char tmpName[8];
+			in.readRawData(tmpName, 8);
+			name = QString(tmpName);
+			file.seek(file.pos() + 12); // dummy
+			in >> offset >> imageConfigSize >> chunks >> zsize >> size;
+			std::vector<int32_t> chunkZsizeArray;
+			for(int j = 0; j < chunks; j++)
+			{
+				int32_t chunkZsize;
+				in >> chunkZsize;
+				chunkZsizeArray.push_back(chunkZsize);
+			}
+			std::vector<int32_t> chunkSizeArray;
+			for(int j = 0; j < chunks; j++)
+			{
+				int32_t chunkSize;
+				in >> chunkSize;
+				chunkSizeArray.push_back(chunkSize);
+			}
+			nameOffset = file.pos();
+
+			file.seek(offset);
+			char tmpImageConfig[100000] = {  };
+			in.readRawData(tmpImageConfig, imageConfigSize);
+			imageConfig = QString(tmpImageConfig);
+			offset += imageConfigSize;
+
+			for(int j = 0; j < chunks; j++)
+			{
+				file.seek(offset);
+				QByteArray tmpData;
+				if(chunkZsizeArray[j] == chunkSizeArray[j])
+				{
+					tmpData.resize(chunkSizeArray.back());
+					in.readRawData(tmpData.data(), chunkSizeArray.back());
+					data += tmpData;
+				}
+				else
+				{
+					tmpData.resize(zsize);
+					in.readRawData(tmpData.data(), zsize);
+					dataCompressed += tmpData;
+				}
+				offset += chunkZsizeArray.back();
+			}
+
+			int dataCompressedLength = dataCompressed.length();
+			while(dataCompressed.length() > 0)
+			{
+				auto tmpData = qUncompress(QByteArray(4, 0) + dataCompressed);
+				if(tmpData.length() == 0)
+					break;
+				dataCompressedArray.push_back(tmpData);
+				int len = qCompress(tmpData).length() - 4; // hack to get length of compressed zlib data
+				dataCompressed.remove(0, len);
+				if(dataCompressed.length() == 0 || dataCompressed.at(0) != 0x78) // check for zlib magic
+					break;
+			}
+
+			if(data.length() < dataCompressedLength)
+				outData = dataCompressedArray;
+			else
+				outData.push_back(data);
+
+			if(cb)
+				cb(name, imageConfig, outData);
+		}
+	}
+}
+
+QImage HdExtractor::loadDds(QByteArray & data)
+{
+	dds::Image image;
+	if(dds::readImage((uint8_t*)data.data(), data.length(), &image) == dds::ReadResult::Success)
+	{
+		assert(image.format == DXGI_FORMAT_BC1_UNORM || image.format == DXGI_FORMAT_BC3_UNORM); // Heroes HD should use only this variant
+		assert(image.mipmaps.size() == 1);
+
+		auto src = image.mipmaps[0].data();
+		uchar *uncompData = new uchar[image.width * image.height * 4];
```

Comment:
what is 4? number of color components? Create a constant.

also, use unique_ptr or the same thing from Qt (QScopedPointer IIRC)